File: src_langgraph/runtime_io.py
from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any
from uuid import UUID, uuid4

from nats import errors as nats_errors
from nats.aio.client import Client as NATS

from src_shared.contracts import (
    ErrorInfo,
    LlmRequest,
    LlmResponse,
    ToolCallRequest,
    ToolCallResponse,
    WorkflowRunRequest,
    now_ts_ms,
)


def _child_trace(parent: WorkflowRunRequest, *, request_id: UUID | None = None) -> dict[str, Any]:
    """Создает trace-блок для дочернего вызова LLM/tool на базе входного запроса."""
    # Preserve trace/correlation across nested LLM/tool calls while generating per-call request_id.
    return {
        "trace_id": parent.trace_id,
        "correlation_id": parent.correlation_id,
        "request_id": request_id or uuid4(),
        "session_id": parent.session_id,
        "ts_ms": now_ts_ms(),
    }


@dataclass(slots=True)
class RuntimeIO:
    """Транспортный слой для request/reply вызовов LLM и tools по NATS."""

    nc: NATS
    llm_subject_prefix: str
    tools_subject_prefix: str
    user_id: str
    request_timeout_s: float = 120.0

    async def call_llm(
        self,
        *,
        parent: WorkflowRunRequest,
        mode: str,
        input_data: dict[str, Any],
        constraints: dict[str, Any] | None = None,
    ) -> LlmResponse:
        """Отправляет запрос в LLM-сервис и валидирует структуру ответа."""
        req = LlmRequest(
            **_child_trace(parent),
            mode=mode,  # type: ignore[arg-type]
            input=input_data,
            constraints=constraints or {},
        )
        subject = f"{self.llm_subject_prefix}{self.user_id}"
        try:
            raw = await self._request_json(subject, req.model_dump_json().encode("utf-8"))
            return LlmResponse.model_validate(raw)
        except nats_errors.TimeoutError:
            return LlmResponse(
                trace_id=req.trace_id,
                correlation_id=req.correlation_id,
                request_id=req.request_id,
                session_id=req.session_id,
                ts_ms=now_ts_ms(),
                ok=False,
                data=None,
                error=ErrorInfo(
                    code="nats_timeout",
                    message=f"Timeout waiting for LLM response from '{subject}'",
                ),
            )
        except Exception as exc:
            return LlmResponse(
                trace_id=req.trace_id,
                correlation_id=req.correlation_id,
                request_id=req.request_id,
                session_id=req.session_id,
                ts_ms=now_ts_ms(),
                ok=False,
                data=None,
                error=ErrorInfo(
                    code="llm_transport_error",
                    message=f"LLM request failed for '{subject}': {exc}",
                ),
            )

    async def call_tool(
        self,
        *,
        parent: WorkflowRunRequest,
        tool_name: str,
        args: dict[str, Any],
    ) -> ToolCallResponse:
        """Вызывает tool-сервис по имени инструмента и валидирует ответ."""
        req = ToolCallRequest(
            **_child_trace(parent),
            tool_name=tool_name,
            args=args,
        )
        subject = f"{self.tools_subject_prefix}{tool_name}"
        try:
            raw = await self._request_json(subject, req.model_dump_json().encode("utf-8"))
            return ToolCallResponse.model_validate(raw)
        except nats_errors.TimeoutError:
            return ToolCallResponse(
                trace_id=req.trace_id,
                correlation_id=req.correlation_id,
                request_id=req.request_id,
                session_id=req.session_id,
                ts_ms=now_ts_ms(),
                ok=False,
                artifact_key=None,
                data=None,
                error=ErrorInfo(
                    code="nats_timeout",
                    message=f"Timeout waiting for tool response from '{subject}'",
                ),
            )
        except Exception as exc:
            return ToolCallResponse(
                trace_id=req.trace_id,
                correlation_id=req.correlation_id,
                request_id=req.request_id,
                session_id=req.session_id,
                ts_ms=now_ts_ms(),
                ok=False,
                artifact_key=None,
                data=None,
                error=ErrorInfo(
                    code="tool_transport_error",
                    message=f"Tool request failed for '{subject}': {exc}",
                ),
            )

    async def _request_json(self, subject: str, payload: bytes) -> dict[str, Any]:
        """Базовый request/reply helper: отправляет payload и требует JSON object в ответе."""
        msg = await self.nc.request(subject, payload, timeout=self.request_timeout_s)
        data = json.loads(msg.data.decode("utf-8"))
        if not isinstance(data, dict):
            raise RuntimeError(f"Invalid JSON response from {subject}")
        return data

File: src_langgraph_rb/lib/src_langgraph_rb/runtime/runtime_io.rb
# frozen_string_literal: true

require "json"
require "nats"

module SrcLanggraphRb
  module Runtime
    class RuntimeIo
      attr_reader :nc, :llm_subject_prefix, :tools_subject_prefix, :user_id, :request_timeout_s

      def initialize(nc:, llm_subject_prefix:, tools_subject_prefix:, user_id:, request_timeout_s: 120.0)
        @nc = nc
        @llm_subject_prefix = llm_subject_prefix
        @tools_subject_prefix = tools_subject_prefix
        @user_id = user_id
        @request_timeout_s = request_timeout_s.to_f
      end

      def call_llm(parent:, mode:, input_data:, constraints: nil)
        req = Contracts::Llm.request(parent:, mode:, input: input_data, constraints: constraints || {})
        subject = "#{llm_subject_prefix}#{user_id}"
        raw = request_json(subject, JSON.generate(Runtime::Util.deep_stringify(req)))
        Contracts::Llm.normalize_response(raw, request: req)
      rescue StandardError => e
        request = req || fallback_request(parent)
        Contracts::Llm.error_response(
          request: request,
          code: "llm_transport_error",
          message: "LLM request failed for '#{subject}': #{e.message}"
        )
      end

      def call_tool(parent:, tool_name:, args:)
        req = Contracts::Tools.request(parent:, tool_name:, args:)
        subject = "#{tools_subject_prefix}#{tool_name}"
        raw = request_json(subject, JSON.generate(Runtime::Util.deep_stringify(req)))
        Contracts::Tools.normalize_response(raw, request: req)
      rescue StandardError => e
        request = req || fallback_request(parent)
        Contracts::Tools.error_response(
          request: request,
          code: "tool_transport_error",
          message: "Tool request failed for '#{subject}': #{e.message}"
        )
      end

      private

      def fallback_request(parent)
        trace = Runtime::Util.extract_hash(parent)
        {
          trace_id: trace[:trace_id] || Runtime::Util.generate_uuid,
          correlation_id: trace[:correlation_id] || trace[:trace_id] || Runtime::Util.generate_uuid,
          request_id: Runtime::Util.generate_uuid,
          session_id: trace[:session_id],
          ts_ms: Contracts::Common.now_ts_ms
        }
      end

      def request_json(subject, payload)
        msg = nc.request(subject, payload, timeout: request_timeout_s)
        data = JSON.parse(msg.data)
        raise ValidationError, "Invalid JSON response from #{subject}" unless data.is_a?(Hash)

        data
      end
    end
  end
end
